### ISSUE TITLE
Add a cattrs like Struct.structure(**kwargs) method

### ISSUE THREAD

**Issue by Goldziher** [jcrist/msgspec#373]
### Description

Hi @jcrist ,

So we previously hit a mutual wall with parsing/validation as part of the Struct init method. I'd like to suggest adding functionality similar to that in cattrs - `structuring` with structuring hooks. This made it very simple for us to integrate this, and it would unblock us with msgspec as well. 

Lemme know if you're game. Cheers

**Comment by Goldziher** [comment]
@jcrist ?

**Comment by jcrist** [comment]
How would this be different than the existing `msgspec.from_builtins`? Is the desired feature the existence of a method on all `Struct` class named `structure`? Or the functionality that this method would provide (which could be done by a function instead of a method on a `Struct` class)? Can you provide a code example showing how this would be used?

**Comment by Goldziher** [comment]
> How would this be different than the existing `msgspec.from_builtins`? Is the desired feature the existence of a method on all `Struct` class named `structure`? Or the functionality that this method would provide (which could be done by a function instead of a method on a `Struct` class)? Can you provide a code example showing how this would be used?

Hi,

Well this is how we use the cattrs converter to take raw kwargs and parse them into a class instance while running the validations - https://github.com/litestar-org/litestar/blob/main/litestar/_signature/models/attrs_signature_model.py#L305

We than "unstructure" them into a python dict. 

The difference is that structuring allow for the conversation of a dict of python builtins (including stdlib stuff like datetime and UUID) and transform them into higher types using configurable hooks. 

Unstructuring is basically the reverse process. 

As for how this could be implemented - a functions similar to `from_buitins` would be great, and I guess also it's the most consistent. But I am really not picky about this stuff.

**Comment by jcrist** [comment]
> The difference is that structuring allow for the conversation of a dict of python builtins (including stdlib stuff like datetime and UUID) and transform them into higher types using configurable hooks.

This is exactly what `from_builtins` does. Can you provide an example of an input/output type that doesn't work the way you'd like with this existing method?

**Comment by Goldziher** [comment]
> > The difference is that structuring allow for the conversation of a dict of python builtins (including stdlib stuff like datetime and UUID) and transform them into higher types using configurable hooks.
> 
> This is exactly what `from_builtins` does. Can you provide an example of an input/output type that doesn't work the way you'd like with this existing method? How is the existing function insufficient?

Hmm... Ok,lemme try again and see. Iirc there was something but i can't recall what. I'll update

**Comment by jcrist** [comment]
`msgspec.from_builtins` was renamed to `msgspec.convert` in the latest release (with some expanded features). Closing this as resolved for now. Please open a new issue if you find any missing features/issues with `msgspec.convert`.
